Spinlock validation augmentation

andyross opened the pull request:
No need to merge this for 1.14.  It's a check I added at one point when we noticed a spot where the kernel was trying to swap away with a spinlock held (recursive irqlocks worked like that, but with spinlocks you must be 100% sure there is only one lock being released in _Swap() -- and that's a good thing!).

It's cheap and easy, but as it happens everything else passes with it, so there's no rush to get it into the tree.

Commit:
kernel/spinlock: Add a depth count to spinlock validation

Right now the validation layer catches mismatched locking, but another
common gotcha is failing to release an outer nested lock before
entering a blocking primitive.  In these cases, the OS would swap
away, and the next thread (if any) to try to take the outer lock would
hit a "not my spinlock" error, which while correct isn't very
informative.

By keeping a single per-cpu nesting count we can check this at the
point of the swap, which is more helpful.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented on kernel/thread.c:
```diff
@@ -715,6 +715,7 @@ int z_spin_lock_valid(struct k_spinlock *l)
 		}
 	}
 	l->thread_cpu = _current_cpu->id | (u32_t)_current;
+	_current_cpu->spin_depth++;
```
needs an #ifdef since spin_depth isn't always there

andrewboie commented on kernel/include/kernel_structs.h:
```diff
@@ -107,6 +108,10 @@ struct _cpu {
 	u8_t swap_ok;
 #endif
+
+#ifdef SPIN_VALIDATE
```
Can we make SPIN_VALIDATE a Kconfig?

andrewboie commented on kernel/thread.c:
```diff
@@ -724,6 +725,7 @@ int z_spin_unlock_valid(struct k_spinlock *l)
 		return 0;
 	}
 	l->thread_cpu = 0;
+	_current_cpu->spin_depth--;
```
needs an #ifdef since spin_depth isn't always there

Commit:
kernel/spinlock: More validation

Catching the error in a function is nice, but one really wants to know
where it happened, and where the recursive lock was taken (or where
the unowned release was actually grabbed).  Add a layer of macro
indirection to catch this info and log it with the assertion.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
